internal/controller/apiextensions/definition/reconciler.go

```diff
@@ -630,3 +631,23 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 
 	return reconcile.Result{Requeue: false}, errors.Wrap(r.client.Status().Update(ctx, d), errUpdateStatus)
 }
+
+// ControllerNeedsRestart returns true if the composite resource controller
+// needs to be restarted based on the XRD's current generation compared to
+// when the controller was last started.
+func ControllerNeedsRestart(d *v1.CompositeResourceDefinition) bool {
+	c := d.GetCondition(v1.TypeEstablished)
+
+	// Only check if we have a WatchingComposite condition
+	if c.Reason != v1.ReasonWatchingComposite {
+		return false
+	}
+
+	// If observedGeneration is 0, the condition was never properly set
+	if c.ObservedGeneration == 0 {
```

Comment:
remind me how this happens - would this be a condition that was set in the past before scott added observedGeneration support in conditions? or are there other reasons? 🤔